### DISCUSSION TITLE
Need help setting up nvim-dap for c++

### DISCUSSION THREAD

**Discussion by BillyData** [mfussenegger/nvim-dap#1054] [Q&A] [answered]
When I run the debugger, the debugger does not stop at the breakpoint, here is my config file, can anyone help me out please ?
https://github.com/BillyData/neovimconfig

**Comment by ochsavidare** [comment]
Did you compile with debug symbols?

**Reply by BillyData** [reply]
Here is how I compile : autocmd filetype cpp nnoremap <F5> :w <bar> !g++ -std=c++17 -O2 -Wall % -o %:r <CR>

Note I have a script that autogenerates CMakeLists.txt and map key_binding to auto generate compile_commands.json so basically when I just run my programm normally it works but when I run the debugger it does not stop

**Reply by theHamsta** [reply]
When you add a `-g` flag, it will compile with debug information

**Reply by ochsavidare** [reply]
Yes, that should make it work. It’s also a good thing to make sure it works when running lldb (if that’s what you’re using) outside of nvim.

**Reply by BillyData** [reply]
![](https://user-images.githubusercontent.com/119795831/274456905-64d93085-7d8b-433b-a0c3-73668d57811c.png)

**Reply by ochsavidare** [reply]
Does it work with that?

**Reply by BillyData** [reply]
It does not, how can I check if I have set up lldb correctly

**Reply by ochsavidare** [reply]
Like I said I think it good to check lldb outside of nvim first, so in  a terminal do something similar to:

>lldb -- a.out 1 2 3
>breakpoint set --file main.c --line 12
>run

then if lldb is working and you have all debug symbols setup correctly, you should break at line 12 in main.c. If not, I think you have to make that work before adding the complexity of the nvim layer.

**Reply by BillyData** [reply]
<img width="828" alt="Screenshot 2023-10-12 at 15 09 35" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/68752384-e457-48b3-82be-6b4693fa3bba">
<img width="1591" alt="Screenshot 2023-10-12 at 15 09 57" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/12778967-ff80-4b5f-a2e6-fac800991596">
![Uploading Screenshot 2023-10-12 at 15.10.24.png…]()

**Reply by BillyData** [reply]
But anyway I set up lldb-vscode in nvim-dap.lua, because I did not find lldb choice in nvim-dap configuration page

**Reply by BillyData** [reply]
But I did downloaded lldb-vscode

**Reply by BillyData** [reply]
Here is the rest of the solution.cpp file, I dont why github does not upload the final screenshot :
<img width="1131" alt="Screenshot 2023-10-12 at 15 19 50" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/a17b6732-ad3d-487d-81da-73be91417208">

**Reply by ochsavidare** [reply]
Yes, sure, lldb is the actual debugger, lldb-vscode is just what nvim uses to communicate with the debugger.

**Reply by ochsavidare** [reply]
Seems like it’s working on its own. That’s good.

**Reply by ochsavidare** [reply]
You could try to increase the log level for dap in nvim, rerun and see what the log is saying. If the beakpoint is accepted and so on.

**Reply by BillyData** [reply]
sorry I am pretty new to programming and especially debugger, how to increase the log level for dap in nvim ? Thanks

**Reply by ochsavidare** [reply]
No problem, we all have to start somewhere :)

do
:DapSetLogLevel TRACE (or whatever level you want)
Rerun and then do
:DapShowLog

**Reply by BillyData** [reply]
Can I ask where are you from and why it seems like you never sleep , I am in Viet Nam right now

**Reply by ochsavidare** [reply]
Haha, I’m from Sweden. :)

**Reply by BillyData** [reply]
<img width="1462" alt="Screenshot 2023-10-12 at 17 21 25" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/33184393-e299-4267-8cb2-0737bbb7a16f">
I cant understand why because when I place the breakpoint there is a B in the line

**Reply by ochsavidare** [reply]
Yeah, the B just means it will try to put a breakpoint there, not that it succeeds. 
So I guess we found the problem (just not the solution). You have both a server disconnect, and a rejected breakpoint. Both not good.

**Reply by BillyData** [reply]
Oh damn, you have checked my config file right ?

**Reply by BillyData** [reply]
<img width="868" alt="Screenshot 2023-10-12 at 17 29 46" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/3d9f6959-cf34-49d9-abfc-a8b4ce3feeca">
Here is also some of what I have downloaded

**Reply by BillyData** [reply]
<img width="807" alt="Screenshot 2023-10-12 at 17 30 36" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/e9486c16-4bc2-4826-b19e-9c7062246023">

**Reply by ochsavidare** [reply]
Are you sure about the path to lldb-vsvode? I know I have mine in /usr/bin/lldb-vscode

**Reply by BillyData** [reply]
<img width="1391" alt="Screenshot 2023-10-12 at 17 31 17" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/3764c58d-d577-4869-87eb-1d6ac15fe6ec">

**Reply by BillyData** [reply]
I will push my whole nvim folder again to Github so you can have the most up to date version of it

**Reply by BillyData** [reply]
Okay I have pushed it up

**Reply by BillyData** [reply]
After I check /usr/bin
<img width="240" alt="Screenshot 2023-10-12 at 17 37 52" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/ab42185d-45ab-4cf1-821d-d95d0b5e976a">

**Reply by BillyData** [reply]
<img width="324" alt="Screenshot 2023-10-12 at 17 39 01" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/1f39ccc1-66e9-4619-9d15-e67c37365380">
After I check /usr/local/bin

**Reply by ochsavidare** [reply]
Yes, I just saw the screenshot with the which command 👍

**Reply by ochsavidare** [reply]
Just curious, how did you install lldb-vscode? I did it through apt-get and then got both lldb and lldb-vscode together

**Reply by BillyData** [reply]
I symlink it from what I remembered

**Reply by BillyData** [reply]
if necessary I can download if again, but can you show me how you did it ?

**Reply by BillyData** [reply]
honestly I have no idea how I have it, maybe I download some extension  on vscode and I think chatgpt or some source tell me to symlink it

**Reply by BillyData** [reply]
but we can do it again, can you show me how to uninstall it and then install it your way

**Reply by ochsavidare** [reply]
Not sure if it will make a difference, just asking random things I’m thinking about 😅

to remove and then install you should be able to do
> sudo apt-get remove lldb -y 
> sudo apt-get update 
> sudo apt-get install lldb

**Reply by ochsavidare** [reply]
I guess maybe remove lldb-vscode also

**Reply by BillyData** [reply]
<img width="298" alt="Screenshot 2023-10-12 at 18 27 37" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/e0ba8b52-5653-41c9-82e9-56ea2f646171">

**Reply by BillyData** [reply]
Anyway I think lldb is built on macos

**Reply by BillyData** [reply]
I have downloaded XCode so it comes along

**Reply by BillyData** [reply]
I think I should only redownload lldb-vscode

**Reply by ochsavidare** [reply]
Ahhh you’re on macOS? I’ve been giving you Linux commands 😬

**Reply by ochsavidare** [reply]
Lldb exists for Linux as well

**Reply by ochsavidare** [reply]
Another idea is just to try codelldb instead since I saw above you installed that as well

**Reply by BillyData** [reply]
yeh but I dont know how to configure it

**Reply by ochsavidare** [reply]
Have you seen the help?
https://github.com/mfussenegger/nvim-dap/wiki/C-C---Rust-(via--codelldb)

**Reply by BillyData** [reply]
Actually I did try to configure it and same situation but I can try it again

**Reply by BillyData** [reply]
I think you and I should take a rest as I have lost too much time on configuring this, later after I have done I would tell you. I really appreciate your effort and dedication to help me out

**Reply by ochsavidare** [reply]
No problem, it wasn’t much effort 🙂
Too bad we couldn’t get it to work. You can debug in the terminal at least.

**Reply by BillyData** [reply]
Yeh I guess I would learn to debug from terminal first, but it is okay, my vscode is fully configured already and I can temporarily use it, when I have time I would give nvim a try again ( which is quite sad since I have almost every thing configured except the debugger )

**Reply by ochsavidare** [reply]
Just a last thought from looking at your config, if you use the native lsp client you don’t need coc. Especially since you seem to use both for clangd.

**Reply by ochsavidare** [reply]
Ok, a last last thought, try to remove the optimization when you compile. Maybe the compiler does some magic to optimize away where you’re trying to break.

**Reply by BillyData** [reply]
Sorry but what do you mean optimization, I have heard someone tell me to do that but I dont know what it means

**Reply by BillyData** [reply]
Yeh true it takes me sometimes to finally set up lsp and I did fail setting up it before and there was a time I tried with coc

**Reply by ochsavidare** [reply, marked as answer]
The -O2 flag sets the optimization level, so removing that

**Reply by BillyData** [reply]
Holy shit ! It runs bro, but now I dont know where to input the argument lol 
<img width="1125" alt="Screenshot 2023-10-13 at 13 27 25" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/8860a799-894d-4734-9566-8fa371d8ba98">

**Reply by BillyData** [reply]
Thanks so much bro, but now how to input :)))

**Reply by ochsavidare** [reply]
What do you mean input the arguments? The optimization flag should only have to do with the compilation and not how you send input to the program?

**Reply by BillyData** [reply]
no i mean i dont know how to input the value for the program

**Reply by BillyData** [reply]
I think I should input in the DAP Console right but when I try to input it says cannot make changes, modifcable is off

**Reply by ochsavidare** [reply]
Ah you mean capture the stdin? I don’t know about that. Might not be supported by lldb-vscode from what I read here:

https://github.com/mfussenegger/nvim-dap/discussions/557

but you could rewrite your program to either read from a file or to accept arguments at launch.

**Reply by BillyData** [reply]
oh okay no problem then, thanks so much for your help but can I ask what is DAP Console is used for ?

**Reply by ochsavidare** [reply]
Anyway, guess you could open another discussion about that, since it’s running now at least :)

**Reply by BillyData** [reply]
:))) okay I will create another discussion about it, now I can finally have a fully functional neovim with so much help from so many kind people like you. Again I really appreciate your help and effort to help me out with such dedication

**Comment by BillyData** [comment]
<img width="1342" alt="Screenshot 2023-10-12 at 08 48 29" src="https://github.com/mfussenegger/nvim-dap/assets/119795831/64d93085-7d8b-433b-a0c3-73668d57811c">

**Reply by BillyData** [reply]
But I did downloaded lldb-vscode
